### PR TITLE
[RSDK-9072] - Simplify stream inits

### PR DESCRIPTION
**Pull request by seanavery**
## Description
Getting started on dynamic resolution epic. This is just a first-step PR to make sure I have a good handle on the relevant code & tests.

- Simplifies stream server initialization.
  - No longer special case for noop stream server.
  - Removes passing in streams to stream server constructor `webstream.NewServer`.
- Unifies logic for adding new streams between reconfigures and initialization.
  - No longer have separate stream flows for `initStreamServer` vs `Reconfigure`.
  - Simplified `addNewStreams` helper.
  - Added `createStream` helper for commonalities between audio & video stream initialization.

## Testing
- webcam live stream ✅ 
  - linux ✅ 
  - darwin ✅ 
  - reconfigure ✅ 
- fake cam live stream ✅ 
  - GetImage ✅ 
  - passthrough ✅ 
    - reconfigure ✅ 
- audio stream ✅
  - reconfigure ✅ 
 - viamrtsp ✅ 
   - passthrough ✅ 
   - GetImage ✅ 
   - reconfigure ✅

### CODE REVIEW

File: robot/web/web_c.go
Comment (seanavery) [inline robot/web/web_c.go:179]: Removed framerate prop passthrough to TargetFrameRate encoders settings for now.
- This was not being checked for reconfigures, only on initialization.
- After testing different cameras, this was only being triggered for Linux webcams.
  - Throttling stream by TargetFrameRate was not even helping as the stream.Next() for webcams waits for the next frame anyhow.
  
I can add this back in f we want, but created a ticket and TODO comment to address this in a more holistic way for all cameras. We will want reliable access to the framerate for bitrate calculations from different dynamic resolution options.

File: robot/web/stream/server.go
Comment (randhid) [inline robot/web/stream/server.go:55]: Since this is a breakign change, can you check if goutils or app is using this function?

File: robot/web/web_c.go
Comment (randhid) [inline robot/web/web_c.go:86]: Why are we returning an error now?

**Review by randhid — COMMENTED** [review]
Can you also check :
- [ ] rtsp camera module with passthrough enabled 
- [ ]  rtsp camera module without passthrough enabled

**Comment by hexbabe** [comment]
Could we include in manual testing a check to verify that there are no audio stream regressions?

**Review by hexbabe — CHANGES_REQUESTED** [review]
Could we include in manual testing a check to verify that there are no audio stream regressions? Testing with fake should be good ya

File: robot/web/web_c.go
Comment (hexbabe) [reply robot/web/web_c.go:179]: This is the only code path we're removing right? All the other changes basically just change the stream registration ownership from `server.go` to `web_c.go?

File: robot/web/web_c.go
Comment (dgottlieb) [inline robot/web/web_c.go:119]: @seanavery can you double check this for me? I'm pretty confident I described how `createStream` behaves, but I wasn't convinced `startVideoStream` coped with that.

File: robot/web/web_c.go
Comment (dgottlieb) [inline robot/web/web_c.go:97]: I bumped this to `Warn`. I don't expect this to be noisy. And if we were hitting this scenario, I expect a user would like to know why they can't stream stuff without having to turn on (global) debug logging.

File: robot/web/web_c.go
Comment (randhid) [inline robot/web/web_c.go:189]: hmmm. Do we know that we can run streams on windows now?

Comment (seanavery) [reply robot/web/web_c.go:189]: Good call - will add Windows check back in.

File: robot/web/web_c.go
Comment (seanavery) [reply robot/web/web_c.go:156]: Agreed! Might as well go all the way here.

File: robot/web/web_c.go
Comment (seanavery) [reply robot/web/web_c.go:97]: Good call. We should not get into this scenario, but if so, we have problems.

**Review by seanavery — COMMENTED** [review]
Thanks for feedback about testing more! Getting panics with configs that include both viamrtsp and other built-in cameras.

I am thinking this is due to race conditions between `initStreamServer` and `Reconfigure` hits. Still Investigating, but will likely need to lock during initialization to avoid this.

```
        /opt/homebrew/Cellar/go/1.23.2/libexec/src/runtime/debug/stack.go:26 +0x64
runtime/debug.PrintStack()
        /opt/homebrew/Cellar/go/1.23.2/libexec/src/runtime/debug/stack.go:18 +0x1c
go.viam.com/utils.PanicCapturingGoWithCallback.func1.1()
        /Users/sean/go/pkg/mod/go.viam.com/utils@v0.1.106/runtime.go:154 +0x40
panic({0x107274c00?, 0x108f69130?})
        /opt/homebrew/Cellar/go/1.23.2/libexec/src/runtime/panic.go:785 +0x124
go.viam.com/rdk/robot/web.(*webService).addNewStreams(0x14000753880, {0x10770f768, 0x140009f2eb0})
        /Users/sean/rdk/robot/web/web_c.go:109 +0x394
go.viam.com/rdk/robot/web.(*webService).Reconfigure(_, {_, _}, _, {{0x0, 0x0}, {{{0x0, 0x0}, {0x0, 0x0}}, ...}, ...})
        /Users/sean/rdk/robot/web/web_c.go:252 +0x12c
go.viam.com/rdk/robot/impl.(*localRobot).updateWeakDependents.func1.1()
        /Users/sean/rdk/robot/impl/local_robot.go:786 +0x104
go.viam.com/utils.PanicCapturingGoWithCallback.func1()
        /Users/sean/go/pkg/mod/go.viam.com/utils@v0.1.106/runtime.go:164 +0x54
created by go.viam.com/utils.PanicCapturingGoWithCallback in goroutine 204
        /Users/sean/go/pkg/mod/go.viam.com/utils@v0.1.106/runtime.go:151 +0x70
2024-10-22T12:25:08.642-0400    ERROR   utils@v0.1.106/runtime.go:155   panic while running function    {"error": "runtime error: invalid memory address or nil pointer dereference"}
```
___

Update 1:
- This was only happening because viamrtsp module was failing because there is no Darwin version in the registry.
- Still divergent behavior where main does not panic in that situation but this branch does. 


___

Update 2:
- From IRL convo with Dan: we were not properly skipping already registered streams. Fixed this and now not seeing panic issue.

File: robot/web/web_c.go
Comment (seanavery) [reply robot/web/web_c.go:121]: Thanks for flagging this comment. `startVideoStream` does not cope with a nil stream. Added back in `alreadyRegistered` check to manually skip -- will update comment here.

File: robot/web/stream/server.go
Comment (seanavery) [reply robot/web/stream/server.go:55]: Update: We checked through goutils and app and there does not seem to be any stream server `NewServer` refs.

File: robot/web/web_c.go
Comment (seanavery) [reply robot/web/web_c.go:189]: Added in windows check for video streams.
Q: should we also skip audio?

File: robot/web/web_c.go
Comment (randhid) [inline robot/web/web_c.go]: remember to remove in final pass

File: robot/web/web_c.go
Comment (randhid) [reply robot/web/web_c.go:189]: [this](https://github.com/viamrobotics/rdk/blob/main/web/server/entrypoint_windows.go) seems to imply that opus runs fine on windows.

Comment (randhid) [reply robot/web/web_c.go:189]: disclaimer I do not know if the implication is the truth.

Comment (randhid) [reply robot/web/web_c.go:189]: the static build file /etc/ld_wrappers.sh also implies that there is no special casing for opus, it includes a statically linked -lx264 but no opus in the static build flags.

Comment (seanavery) [reply robot/web/web_c.go:189]: Nice find! 
Hm,  yeah I would be surprised if audio drivers worked inside WSL but we can go with it for now.

File: robot/web/web_c.go
Comment (seanavery) [reply robot/web/web_c.go:179]: >  change the stream registration ownership from server.go to web_c.go?
This does not change ownership here. Step 2 after this PR is in will be moving video and audio sources into from `web_c.go` to `server.go`.

File: robot/web/web_c.go
Comment (seanavery) [reply robot/web/web_c.go:88]: Good call. Moved to warning log and returning nil.

File: robot/web/stream/server.go
Comment (randhid) [reply robot/web/stream/server.go:56]: Also before you merge this please post in breaking changes in case someone is psinnig up servers in a module or something creative like that.

File: robot/web/web_c.go
Comment (seanavery) [reply robot/web/web_c.go]: Ok, removed `makeStreamServer`. Punted on fully removing the `StreamServer` wrapper until I fully understand the consequences of `PreferWebRTC` web option.
Created ticket: https://viam.atlassian.net/browse/RSDK-9115

**Review by randhid — APPROVED** [review]
Honestly lgtm if we're all okay with the return from `NewServer` no longer returning an error, which it never did in the past cc @hexbabe @dgottlieb

File: robot/web/stream/server.go
Comment (dgottlieb) [reply robot/web/stream/server.go:56]: > Honestly lgtm if we're all okay with the return from NewServer no longer returning an error, which it never did in the past cc @hexbabe @dgottlieb


This is definitely not part of our "rdk as a library public API", so I'm not worried. But certainly no harm in overcommunicating this as a breaking change.

**Review by dgottlieb — APPROVED** [review]
I love this patch, thanks for taking the initiative to simplifyt his down.

**Review by hexbabe — APPROVED** [review]

**Pull request merged**
